File: pyround1/203.移除链表元素.py
#
# @lc app=leetcode.cn id=203 lang=python
#
# [203] 移除链表元素
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        ans = ListNode(0)
        ans.next = head
        l = ans
        r = ans.next

        while r:
            if r.val == val:
                l.next = r.next
            else:
                l = l.next
            r = r.next
        return ans.next

# @lc code=end

File: 203.移除链表元素.cpp
/*
 * @lc app=leetcode.cn id=203 lang=cpp
 *
 * [203] 移除链表元素
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* node = new ListNode(-1);
        ListNode* res = node;
        node->next = head;
        while(node and node->next){
            while(node->next and node->next->val == val){
                if(node->next->next)
                    node->next = node->next->next;
                else{
                    node->next = nullptr;
                }
            }
            node = node->next;
        }
        return res->next;

    }
};
// @lc code=end

